applicationset/controllers/applicationset_controller.go

```diff
@@ -938,6 +939,21 @@ func (r *ApplicationSetReconciler) buildAppDependencyList(logCtx *log.Entry, app
 	return appDependencyList, appStepMap
 }
 
+func getMatchExpressionValues(matchExpression argov1alpha1.ApplicationMatchExpression) ([]string, error) {
+	if len(matchExpression.ValuesString) > 0 && matchExpression.ValuesString != "" {
+		return nil, errors.New("only one of 'values' or 'valuesString' must be set")
+	}
+	if matchExpression.ValuesString != "" {
+		values := strings.Split(matchExpression.ValuesString, ",")
+		return values, nil
+	}
+	if len(matchExpression.Values) > 0 {
+		return matchExpression.Values, nil
+	}
+
+	return nil, errors.New("nor of the 'values' and 'valuesString' fields are defined")
+}
+
 func labelMatchedExpression(logCtx *log.Entry, val string, matchExpression argov1alpha1.ApplicationMatchExpression) bool {
```

Comment:
I don't think that's the right place.
It should be where it's used as a generator selector:
* cluster selector: https://github.com/argoproj/argo-cd/blob/master/applicationset/generators/cluster.go#L59
* Post selector on all generators: https://github.com/argoproj/argo-cd/blob/master/applicationset/generators/generator_spec_processor.go#L34

Reply:
I don't really get why we should move the function out of this file. It is only used by this function https://github.com/argoproj/argo-cd/blob/b90d673f18abb06551d942b4f5ff166451ac17c5/applicationset/controllers/applicationset_controller.go#L959
here:
https://github.com/argoproj/argo-cd/blob/b90d673f18abb06551d942b4f5ff166451ac17c5/applicationset/controllers/applicationset_controller.go#L970

Which is in the same file. We want the rest of the code to use `Values` and not `ValuesString`. That is why, we do the conversion as soon as possible

Reply:
It's not the same purpose. My proposal was for the applicationset generators selectors. Here it's for the application selectors 😁